plugins/destination/postgresql/client/migrate.go

```diff
@@ -3,43 +3,125 @@ package client
+ CASE 
+		 WHEN contype = 'p' THEN true
+		 ELSE false
+ END AS is_primary_key,
+ CASE 
+		 WHEN pg_attribute.attnotnull THEN true
+		 ELSE false
+ END AS not_null,
+ COALESCE(pg_constraint.conname, '') AS primary_key_constraint_name
+ FROM
+ pg_catalog.pg_attribute
+ INNER JOIN
+ pg_catalog.pg_class ON pg_class.oid = pg_attribute.attrelid
+ INNER JOIN
+ pg_catalog.pg_namespace ON pg_namespace.oid = pg_class.relnamespace
+ LEFT JOIN
+ pg_catalog.pg_constraint ON pg_constraint.conrelid = pg_attribute.attrelid
+ AND pg_constraint.conkey[1] = pg_attribute.attnum
+ AND contype = 'p'
+ INNER JOIN
+ information_schema.columns ON columns.table_name = pg_class.relname AND columns.column_name = pg_attribute.attname AND columns.table_schema = pg_catalog.pg_namespace.nspname
+ WHERE
+ pg_attribute.attnum > 0
+ AND NOT pg_attribute.attisdropped
+ AND pg_catalog.pg_namespace.nspname = '%s'
+ ORDER BY
+ table_name ASC , ordinal_position ASC;
 `
 )
 
+func (c *Client) listPgTables(ctx context.Context, pluginTables schema.Tables) (schema.Tables, error) {
+	var tables schema.Tables
+	rows, err := c.conn.Query(ctx, fmt.Sprintf(selectAllTables, c.currentSchemaName))
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var ordinalPosition int
+		var tableName, columnName, columnType, pkName string
+		var isPrimaryKey, notNull bool
+		if err := rows.Scan(&ordinalPosition, &tableName, &columnName, &columnType, &isPrimaryKey, &notNull, &pkName); err != nil {
+			return nil, err
+		}
+		// This is a small optimization to hold only the tables that are in the plugin
+		if pluginTables.Get(tableName) == nil {
+			continue
+		}
+		if ordinalPosition == 1 {
+			tables = append(tables, &schema.Table{
+				Name:    tableName,
+				Columns: make([]schema.Column, 0),
+			})
+		}
+		table := tables[len(tables)-1]
+		if pkName != "" {
+			table.PkConstraintName = pkName
+		}
+		table.Columns = append(table.Columns, schema.Column{
+			Name: columnName,
+			CreationOptions: schema.ColumnCreationOptions{
+				PrimaryKey: isPrimaryKey,
+				NotNull:    notNull,
+			},
+			Type: c.PgToSchemaType(columnType),
+		})
+	}
+	return tables, nil
+}
+
 // This is the responsibility of the CLI of the client to lock before running migration
 func (c *Client) Migrate(ctx context.Context, tables schema.Tables) error {
-	for _, table := range tables {
+	pgTables, err := c.listPgTables(ctx, tables)
+	if err != nil {
+		return fmt.Errorf("failed listing postgres tables: %w", err)
+	}
+	for _, table := range tables.FlattenTables() {
 		c.logger.Info().Str("table", table.Name).Msg("Migrating table")
 		if len(table.Columns) == 0 {
 			c.logger.Info().Str("table", table.Name).Msg("Table with no columns, skipping")
 			continue
 		}
-		tableExist, err := c.isTableExistSQL(ctx, table.Name)
-		if err != nil {
-			return fmt.Errorf("failed to check if table %s exists: %w", table.Name, err)
+
+		// In postgres if column is primary key, it can't be null
+		if c.enabledPks() {
```

Comment:
what about append mode (`_cq_id` must be marked `unique not null`)?

Reply:
resolved in https://github.com/cloudquery/cloudquery/pull/7819#issuecomment-1428288001